designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+- Feature Name: static_matrices
+- Start Date: April 30th, 2020
+- RFC PR: [Example Branch](https://github.com/stan-dev/math/compare/feature/var-template)
+- Stan Issue: [#1805](https://github.com/stan-dev/math/issues/1805)
+
+# Summary
+[summary]: #summary
+
+This proposes a `static_matrix` and `static_vector` type in the Stan language which on the backend will allow for significant performance opportunities. In Stan Math this will be represented by `var<Eigen::Matrix<double, 1, 1>>` or `var<matrix_cl<double>>`. The implementation will be fully backwards compatible with current Stan code. With optimizations from the compiler the conditions for a `matrix` to be a `static_matrix` can be detected and applied to Stan programs automatically for users.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, a differentiable matrix in Stan is represented as an Eigen matrix holding a pointer to an underlying array of autodiff objects, each of those holding a pointer to the underlying autodiff object implementation. These `N*M` autodiff elements are expensive but necessary so the elements of a matrix can be assigned to without copying the entire matrix. However, in instances where the matrix is treated as a whole object such that underlying subsets of the matrix are never assigned to, we can store one autodiff object representing the entire matrix. 
```

Comment:
Before we finalize the static matrix design doc I think we should discuss a bit the part of the optimizations that also apply to non-static matrix. Namely what I suggested here: https://github.com/stan-dev/math/issues/1805#issuecomment-605836282

Reply:
Do you mean

> I think the first suggestion could be simplified a bit by writing a specialization for Eigen::Matrix<var>. Also we can make non-static version with second member Eigen::Matrix<vari*, -1, -1> adj_;. This is not as efficient as what you suggested, but can be modified.

I can dig the route of a specialization for `Eigen::Matrix<var>`. Though what I'm confused about is, if someone wants to dynamically assign parts of a matrix why wouldn't they just use the current matrix type?

Reply:
@t4c1 if we add `Eigen::Matrix<var>` what would the internal Eigen representation be? And would we have to write specializations for other Eigen types? I kind of like the `var_t<T>` approach since that would also work for sparse matrices (and probably with some other stuff matrix_cl<>`)

Reply:
> I can dig the route of a specialization for Eigen::Matrix<var>

Sorry, that was not the part I wanted to point out here.

> if someone wants to dynamically assign parts of a matrix why wouldn't they just use the current matrix type?

I think we may be able to remake it in a way that would be as efficient as static matrix you are proposing here.

> if we add Eigen::Matrix<var> what would the internal Eigen representation be? 

It would not actually be Eigen type. I was just thinking we could specialize it this way and still implement it the same way as you propose for `var<MatrixXd>` to allow us to keep most functions accepting it the same. Am not sure anymore if that was such a great idea.

> Also we can make non-static version with second member `Eigen::Matrix<vari*, -1, -1> adj_`

That was the part I actually wanted to point out. Although after some more thinking I am not so sure whether this is the best way. 

Maybe we could instead implement changing a part of what you are suggesting as a static matrix to have it completely replace `Matrix<var>`? I think we would have to change some bits of how chain stack works ... 

Just some brainstorming. I need to think about this some more.

Reply:
> I think we may be able to remake it in a way that would be as efficient as static matrix you are proposing here

Oh like replacing the definition of `matrix[N, M]` in the language? imo I think I'd rather leave that to another proposal. Though if it's something simple then I'd be for adding it here. I was kind of confused by the dynamic case you and Bob were talking about. Would we be able to do something simple like a specialization of `var_type`?

Reply:
> Oh like replacing the definition of matrix[N, M] in the language?

No. Just replacing `Eigen::Matrix<var>` in Stan Math. Language needs not to change. 

> imo I think I'd rather leave that to another proposal.

Sure. I was just thinking that if we can do that there is no need for changes in language in this design doc.

Reply:
@t4c1 --- I didn't undestand the `Matrix<vari*, -1, -1>` plan.  That's equivalent to `Matrix<var, -1, -1>` in terms of memory and how it executes (`var` has no virtual methods and a single member `vari*`, so it's memory-equivalent to `vari*`).

There's no way to completely replace `Matrix<var, -1, -1>` with fixed matrices without giving up the ability to assign (that's what I believe JAX and PyTorch do---give up on matrix assignment).  As soon as you assign to an element, you need a much more flexible data structure that loses memory locality.   Now we do want to be able to assign with something like a comprehension or function (say for a covariance matrix for a GP).

Reply:
> I didn't undestand the Matrix<vari*, -1, -1> plan.

It was just an idea. Not the best one for exactley the reasons you are pointing out.

> There's no way to completely replace Matrix<var, -1, -1> with fixed matrices without giving up the ability to assign

Maybe we don't have to give it up. I haven't made any prototypes, so I'm not 100% sure, but I think we may be able to implement assignment for the proposed change to the matrix. It would be more complicated and slower than the assignment currently is - but we are getting speedups elswhere anyway. So if it it turns out it is possible, there is no reason for any changes to the language - we just replace the matrix in Stan Math.